File: String/76-minWindow/76.py
class Solution:
    def cover(self, t_cnt: dict, window: dict) -> bool:
        for k, v in t_cnt.items():
            if window.get(k, 0) < v:
                return False
        return True

    def minWindow(self, s: str, t: str) -> str:
        t_cnt = {c: t.count(c) for c in t}
        window = dict()
        min_len = 100001
        start = -1
        l, r = 0, 0
        for r in range(len(s)):
            if s[r] in t_cnt.keys():
                window[s[r]] = window.get(s[r], 0) + 1
            while self.cover(t_cnt, window) and l <= r:
                cur_len = r - l + 1
                if cur_len < min_len:
                    min_len = cur_len
                    start = l
                window[s[l]] = window.get(s[l], 0) - 1
                l += 1
        return "" if start == -1 else s[start : start + min_len]


if __name__ == "__main__":
    s = "ADOBECODEBANC"
    t = "ABC"
    obj = Solution()
    print(obj.minWindow(s, t))

File: String/76-minWindow/76.cpp
#include <iostream>
#include <limits>
#include <unordered_map>

using namespace std;

class Solution {
 public:
  unordered_map<char, int> window;
  unordered_map<char, int> t_cnt;  // char in t -> count of the char

  bool cover() {
    for (const auto& kv : t_cnt) {
      if (window[kv.first] < kv.second) {
        return false;
      }
    }
    return true;
  }

  string minWindow(string s, string t) {
    for (const char& c : t) {
      ++t_cnt[c];
    }

    int min_len = numeric_limits<int>::max();
    int start = -1;
    for (int l = 0, r = 0; r < int(s.size()); ++r) {
      if (t_cnt.count(s[r])) {
        ++window[s[r]];
      }

      while (cover() && l <= r) {
        int cur_len = r - l + 1;
        if (cur_len < min_len) {
          min_len = cur_len;
          start = l;
        }
        --window[s[l]];
        ++l;
      }
    }

    return start == -1 ? "" : s.substr(start, min_len);
  }
};

int main() {
  string s = "ADOBECODEBANC";
  string t = "ABC";
  Solution obj;
  cout << obj.minWindow(s, t) << endl;

  return 0;
}